sql/core/src/main/scala/org/apache/spark/sql/Melt.scala

```diff
@@ -0,0 +1,143 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql
+
+import org.apache.spark.sql.catalyst.analysis.{AnsiTypeCoercion, TypeCoercion, UnresolvedAttribute}
+import org.apache.spark.sql.catalyst.expressions.{AttributeReference, Cast, Expression, Literal, NamedExpression}
+import org.apache.spark.sql.catalyst.expressions.objects.AssertNotNull
+import org.apache.spark.sql.catalyst.plans.logical.Expand
+import org.apache.spark.sql.functions.col
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types.{DataType, StringType}
+
+private[sql] object Melt {
+  def of[_](
+      ds: Dataset[_],
+      ids: Seq[String],
+      values: Seq[String],
+      dropNulls: Boolean,
+      variableColumnName: String,
+      valueColumnName: String): DataFrame = {
```

Comment:
This function is hard to be reused when we add the SQL syntax for melt. Ideally we need to create a new logical plan like `UnresolvedMelt`, then we add a new analyzer rule to resolve it and rewrite it to existing operators like `Expand`.

For example, the logical plan can use `UnresolvedAttribute` for id/value columns, and leverage the existing resolution framework to resolve them, which takes care of case sensitivity and quoting (e.g. the column name can be `` `id@1` ``). These are all missing in the current imlementation.

Reply:
`Melt` is now a `LogicalPlan` with some rules in the analyser to melt all value columns except for the id columns if no value columns are given. The `Melt` logical plan is replaced by the equivalent `Expand` operation during analysis, once ids columns, value columns and output value type are resolved.

This is much cleaner now, and allows for expressions, structs and `*` for id and value columns. Thanks for the pointers!